File: test/jvm/com/cattailsw/nanidroid/runtime/SakuraScriptPresentationReducerTest.java
package com.cattailsw.nanidroid.runtime;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

import org.junit.Test;

/** Locks the state transitions that will replace SScriptRunner's mutable fields. */
public final class SakuraScriptPresentationReducerTest {
    @Test
    public void requiredMigrationInvariant_scriptResetKeepsSurfacesButClearsTransientPresentation() {
        SakuraScriptPresentationState changed = SakuraScriptPresentationReducer.queueAnimation(
                SakuraScriptPresentationReducer.changeBalloon(
                        SakuraScriptPresentationReducer.changeSurface(
                                SakuraScriptPresentationReducer.selectSpeaker(
                                        SakuraScriptPresentationReducer.initial(), GhostSpeaker.KERO),
                                "42"),
                        "7"),
                "3");

        SakuraScriptPresentationState reset =
                SakuraScriptPresentationReducer.resetForNextScript(changed);

        assertEquals(GhostSpeaker.SAKURA, reset.getActiveSpeaker());
        assertEquals("0", reset.getSakuraSurfaceId());
        assertEquals("42", reset.getKeroSurfaceId());
        assertEquals("-1", reset.getSakuraBalloonId());
        assertEquals("-1", reset.getKeroBalloonId());
        assertNull(reset.getSakuraAnimationId());
        assertNull(reset.getKeroAnimationId());
    }

    @Test
    public void requiredMigrationInvariant_synchronizationAndKeroTextPreserveLegacyBalloonPolicy() {
        SakuraScriptPresentationState state = SakuraScriptPresentationReducer.toggleSynchronization(
                SakuraScriptPresentationReducer.resetForNextScript(
                        SakuraScriptPresentationReducer.initial()));
        state = SakuraScriptPresentationReducer.append(state, 'A');
        GhostPresentationState snapshot = SakuraScriptPresentationReducer.snapshot(state);

        assertEquals("A", snapshot.getSakura().getText());
        assertEquals("A", snapshot.getKero().getText());
        assertTrue(snapshot.getKero().getBalloonVisible());

        state = SakuraScriptPresentationReducer.clearActiveText(state);
        snapshot = SakuraScriptPresentationReducer.snapshot(state);
        assertFalse(snapshot.getSakura().getBalloonVisible());
        assertTrue(snapshot.getKero().getBalloonVisible());
    }

    @Test
    public void requiredMigrationInvariant_animationIsVisibleOnceThenExplicitlyConsumed() {
        SakuraScriptPresentationState queued = SakuraScriptPresentationReducer.queueAnimation(
                SakuraScriptPresentationReducer.initial(), "3");

        assertEquals("3", SakuraScriptPresentationReducer.snapshot(queued)
                .getSakura().getAnimationId());
        assertNull(SakuraScriptPresentationReducer.snapshot(
                SakuraScriptPresentationReducer.consumeAnimations(queued))
                .getSakura().getAnimationId());
    }
}

File: src/test/java/com/cattailsw/nanidroid/runtime/SakuraScriptPresentationReducerTest.kt
package com.cattailsw.nanidroid.runtime

import org.junit.Assert
import org.junit.Test

/** Locks the state transitions that will replace SScriptRunner's mutable fields.  */
class SakuraScriptPresentationReducerTest {
    @Test
    fun requiredMigrationInvariant_scriptResetKeepsSurfacesButClearsTransientPresentation() {
        val changed =
            SakuraScriptPresentationReducer.queueAnimation(
                SakuraScriptPresentationReducer.changeBalloon(
                    SakuraScriptPresentationReducer.changeSurface(
                        SakuraScriptPresentationReducer.selectSpeaker(
                            SakuraScriptPresentationReducer.initial(), GhostSpeaker.KERO
                        ),
                        "42"
                    ),
                    "7"
                ),
                "3"
            )

        val reset = SakuraScriptPresentationReducer.resetForNextScript(changed)

        Assert.assertEquals(GhostSpeaker.SAKURA, reset.activeSpeaker)
        Assert.assertEquals("0", reset.sakuraSurfaceId)
        Assert.assertEquals("42", reset.keroSurfaceId)
        Assert.assertEquals("-1", reset.sakuraBalloonId)
        Assert.assertEquals("-1", reset.keroBalloonId)
        Assert.assertNull(reset.sakuraAnimationId)
        Assert.assertNull(reset.keroAnimationId)
    }

    @Test
    fun requiredMigrationInvariant_synchronizationAndKeroTextPreserveLegacyBalloonPolicy() {
        var state =
            SakuraScriptPresentationReducer.toggleSynchronization(
                SakuraScriptPresentationReducer.resetForNextScript(
                    SakuraScriptPresentationReducer.initial()
                )
            )
        state = SakuraScriptPresentationReducer.append(state, 'A')
        var snapshot = SakuraScriptPresentationReducer.snapshot(state)

        Assert.assertEquals("A", snapshot.sakura.text)
        Assert.assertEquals("A", snapshot.kero.text)
        Assert.assertTrue(snapshot.kero.balloonVisible)

        state = SakuraScriptPresentationReducer.clearActiveText(state)
        snapshot = SakuraScriptPresentationReducer.snapshot(state)
        Assert.assertFalse(snapshot.sakura.balloonVisible)
        Assert.assertTrue(snapshot.kero.balloonVisible)
    }

    @Test
    fun requiredMigrationInvariant_reselectingCurrentSpeakerRetainsItsText() {
        var state = SakuraScriptPresentationReducer.resetForNextScript(
            SakuraScriptPresentationReducer.initial()
        )
        state = SakuraScriptPresentationReducer.append(state, 'A')
        state = SakuraScriptPresentationReducer.selectSpeaker(state, GhostSpeaker.SAKURA)
        state = SakuraScriptPresentationReducer.append(state, 'B')

        Assert.assertEquals("AB", state.sakuraText)

        state = SakuraScriptPresentationReducer.selectSpeaker(state, GhostSpeaker.KERO)
        state = SakuraScriptPresentationReducer.append(state, 'C')
        state = SakuraScriptPresentationReducer.selectSpeaker(state, GhostSpeaker.SAKURA)
        Assert.assertEquals("", state.sakuraText)
        Assert.assertEquals("C", state.keroText)
    }

    @Test
    fun requiredMigrationInvariant_animationIsVisibleOnceThenExplicitlyConsumed() {
        val queued = SakuraScriptPresentationReducer.queueAnimation(
            SakuraScriptPresentationReducer.initial(), "3"
        )

        Assert.assertEquals(
            "3", SakuraScriptPresentationReducer.snapshot(queued).sakura.animationId
        )
        Assert.assertNull(
            SakuraScriptPresentationReducer.snapshot(
                SakuraScriptPresentationReducer.consumeAnimations(queued)
            ).sakura.animationId
        )
    }
}